File: src/Catalog.API/Model/BundleDeal.cs
using System.Text.Json.Serialization;

namespace Chillax.Catalog.API.Model;

/// <summary>
/// Represents a bundle deal (combo) offering multiple items at a special price
/// </summary>
public class BundleDeal
{
    public int Id { get; set; }

    /// <summary>
    /// Localized name of the bundle
    /// </summary>
    public LocalizedText Name { get; set; } = new();

    /// <summary>
    /// Localized description of the bundle
    /// </summary>
    public LocalizedText Description { get; set; } = new();

    /// <summary>
    /// The special bundle price
    /// </summary>
    public decimal BundlePrice { get; set; }

    /// <summary>
    /// Optional picture file name
    /// </summary>
    public string? PictureFileName { get; set; }

    /// <summary>
    /// Whether this bundle is currently active and visible to customers
    /// </summary>
    public bool IsActive { get; set; }

    /// <summary>
    /// Display order (lower numbers appear first)
    /// </summary>
    public int DisplayOrder { get; set; }

    /// <summary>
    /// Items included in this bundle
    /// </summary>
    public ICollection<BundleDealItem> Items { get; set; } = new List<BundleDealItem>();

    /// <summary>
    /// Required for EF Core
    /// </summary>
    private BundleDeal() { }

    [JsonConstructor]
    public BundleDeal(LocalizedText name, LocalizedText? description = null)
    {
        Name = name;
        Description = description ?? new LocalizedText(string.Empty);
    }
}

File: src/admin_app/lib/features/menu/models/bundle_deal.dart
import '../../../core/models/localized_text.dart';

/// Bundle deal model for admin
class BundleDeal {
  final int id;
  final LocalizedText name;
  final LocalizedText description;
  final double bundlePrice;
  final double originalPrice;
  final String? pictureUri;
  final bool isActive;
  final int displayOrder;
  final List<BundleDealItem> items;

  BundleDeal({
    required this.id,
    required this.name,
    required this.description,
    required this.bundlePrice,
    required this.originalPrice,
    this.pictureUri,
    this.isActive = true,
    this.displayOrder = 0,
    this.items = const [],
  });

  double get savings => originalPrice - bundlePrice;

  factory BundleDeal.fromJson(Map<String, dynamic> json) {
    return BundleDeal(
      id: json['id'] as int,
      name: LocalizedText.parse(json['name']),
      description: LocalizedText.parse(json['description'] ?? ''),
      bundlePrice: (json['bundlePrice'] as num).toDouble(),
      originalPrice: (json['originalPrice'] as num).toDouble(),
      pictureUri: json['pictureUri'] as String?,
      isActive: json['isActive'] as bool? ?? true,
      displayOrder: json['displayOrder'] as int? ?? 0,
      items: (json['items'] as List<dynamic>?)
              ?.map((e) => BundleDealItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toRequestJson() {
    return {
      'name': name.toJson(),
      'description': description.toJson(),
      'bundlePrice': bundlePrice,
      'isActive': isActive,
      'displayOrder': displayOrder,
      'items': items.map((e) => {
        'catalogItemId': e.catalogItemId,
        'quantity': e.quantity,
      }).toList(),
    };
  }
}

/// Item within a bundle deal
class BundleDealItem {
  final int id;
  final int catalogItemId;
  final LocalizedText itemName;
  final double itemPrice;
  final int quantity;

  BundleDealItem({
    required this.id,
    required this.catalogItemId,
    required this.itemName,
    required this.itemPrice,
    this.quantity = 1,
  });

  factory BundleDealItem.fromJson(Map<String, dynamic> json) {
    return BundleDealItem(
      id: json['id'] as int,
      catalogItemId: json['catalogItemId'] as int,
      itemName: LocalizedText.parse(json['itemName'] ?? ''),
      itemPrice: (json['itemPrice'] as num).toDouble(),
      quantity: json['quantity'] as int? ?? 1,
    );
  }
}
